### ISSUE TITLE
[JENKINS-64778] Default container selector for the template configuration in the UI

### ISSUE THREAD

**Issue by jenkins-infra-bot** [jenkinsci/kubernetes-plugin#2427]
<p>This is a request to expose the existing <tt>defaultContainer</tt> option (as pointed below) to the Cloud configuration, in the Kubernetes Template side.</p>

<pre>
 agent {
    kubernetes {
      defaultContainer <span class="code-quote">'maven'</span>
      yamlFile <span class="code-quote">'KubernetesPod.yaml'</span>
    }
  }
</pre>

<p>This is how the new UI would be expected to be:</p>

<p> <span class="image-wrap" style=""><a id="53922_thumb" href="https://issues.jenkins.io/secure/attachment/53922/53922_example-of-ui.png" title="example-of-ui.png" file-preview-type="image" file-preview-id="53922" file-preview-title="example-of-ui.png"><img src="https://issues.jenkins.io/secure/thumbnail/53922/_thumb_53922.png" style="border: 0px solid black" role="presentation"/></a></span></p>

---
<details><summary><i>Originally reported by <img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/felipecassiors.png" title="felipecassiors's avatar" /> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=felipecassiors">felipecassiors</a>, imported from: <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-64778" target="_blank">Default container selector for the template configuration in the UI</a></i></summary>
<i><ul>
<li><b>status</b>: Resolved
<li><b>priority</b>: Minor
<li><b>component(s)</b>: kubernetes-plugin
<li><b>resolution</b>: Won't Do
<li><b>resolved</b>: 2021-03-18T22:48:56+00:00
<li><b>votes</b>: 0
<li><b>watchers</b>: 4
<li><b>imported</b>: 2025-12-02
</ul></i>
<details><summary>Raw content of original issue</summary>

<pre>
<p>This is a request to expose the existing <tt>defaultContainer</tt> option (as pointed below) to the Cloud configuration, in the Kubernetes Template side.</p>

<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-yaml"> agent {
    kubernetes {
      defaultContainer <span class="code-quote">'maven'</span>
      yamlFile <span class="code-quote">'KubernetesPod.yaml'</span>
    }
  }
</pre>
</div></div>

<p>This is how the new UI would be expected to be:</p>

<p> <span class="image-wrap" style=""><a id="53922_thumb" href="https://issues.jenkins.io/secure/attachment/53922/53922_example-of-ui.png" title="example-of-ui.png" file-preview-type="image" file-preview-id="53922" file-preview-title="example-of-ui.png"><img src="https://issues.jenkins.io/secure/thumbnail/53922/_thumb_53922.png" style="border: 0px solid black" role="presentation"/></a></span></p></pre>
</details>
</details>
<details><summary><i>1 attachment</i></summary>

- [example-of-ui.png](https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-last/refs/heads/main/attachments/53922/example-of-ui.png)
  > ![example-of-ui.png](https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-last/refs/heads/main/attachments/53922/example-of-ui.png)
</details>

**Comment by jenkins-infra-bot** [comment]
- _Remote link associated with this issue:_


  - [felipecrs/jenkins-agent-dind PR #​16](https://github.com/felipecrs/jenkins-agent-dind/pull/16)

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/felipecassiors.png" title="felipecassiors's avatar" /> felipecassiors:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-64778?focusedId=404204&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-404204">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>This seems to be the only way to choose what container to use: being the jnlp container.</p>

<p>The drawback is that now I have an overloaded container with the Docker Daemon running in parallel with the JNLP service.</p>
</pre>
</details>

</ul>
</details>

<p>This seems to be the only way to choose what container to use: being the jnlp container.</p>

<p>The drawback is that now I have an overloaded container with the Docker Daemon running in parallel with the JNLP service.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/gsanchezu.png" title="gsanchezu's avatar" /> gsanchezu:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-64778?focusedId=404278&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-404278">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Hi <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=felipecassiors" class="user-hover" rel="felipecassiors">felipecassiors</a>, we are no longer actively maintaining this plugin. However, if you find it useful and want to add the suggested functionality, feel free to send a Pull Request for it.</p>

<p> </p>

<p>Thanks!</p>
</pre>
</details>

</ul>
</details>

<p>Hi <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=felipecassiors" class="user-hover" rel="felipecassiors">felipecassiors</a>, we are no longer actively maintaining this plugin. However, if you find it useful and want to add the suggested functionality, feel free to send a Pull Request for it.</p>

<p> </p>

<p>Thanks!</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jglick.png" title="jglick's avatar" /> jglick:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-64778?focusedId=406392&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-406392">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I do not think this is possible to implement. In order to use a nondefault container, the <tt>container</tt> step must be called. In Declarative Pipeline, that is normally done by using <tt>agent kubernetes</tt>. If you just use <tt>agent label</tt> (or run a non-Pipeline build), the <tt>kubernetes</tt> plugin is not even consulted.</p>
</pre>
</details>

</ul>
</details>

<p>I do not think this is possible to implement. In order to use a nondefault container, the <tt>container</tt> step must be called. In Declarative Pipeline, that is normally done by using <tt>agent kubernetes</tt>. If you just use <tt>agent label</tt> (or run a non-Pipeline build), the <tt>kubernetes</tt> plugin is not even consulted.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/felipecassiors.png" title="felipecassiors's avatar" /> felipecassiors:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-64778?focusedId=406394&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-406394">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Well, currently `jnlp` seems to be hardcoded as the default container. The same way `jnlp` is used if none is specified, this option should provide the user the ability to select the default container rather than using `jnlp`.</p>
</pre>
</details>

</ul>
</details>

<p>Well, currently `jnlp` seems to be hardcoded as the default container. The same way `jnlp` is used if none is specified, this option should provide the user the ability to select the default container rather than using `jnlp`.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jglick.png" title="jglick's avatar" /> jglick:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-64778?focusedId=406396&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-406396">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>That is not possible. Using the <tt>jnlp</tt> container is qualitatively different from using other containers via the <tt>container</tt> step.</p>
</pre>
</details>

</ul>
</details>

<p>That is not possible. Using the <tt>jnlp</tt> container is qualitatively different from using other containers via the <tt>container</tt> step.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>jack_zhang:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-64778?focusedId=448793&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-448793">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Here is a workaround for your reference</p>

<p><a href="https://github.com/zhan9san/init-jnlp" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/zhan9san/init-jnlp</a></p>
</pre>
</details>

</ul>
</details>

<p>Here is a workaround for your reference</p>

<p><a href="https://github.com/zhan9san/init-jnlp" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/zhan9san/init-jnlp</a></p>

**Comment by jenkins-infra-bot** [comment]
<i>[Original `relates_to` from Jira: <a href="https://github.com/jenkinsci/kubernetes-plugin/issues?q=is%3Aissue%20%22jira_issue_key%3DJENKINS-41109%22">JENKINS-41109</a>]</i>

**Comment by jenkins-infra-bot** [comment]
<i>[Original `relates_to` from Jira: <a href="https://github.com/jenkinsci/kubernetes-plugin/issues?q=is%3Aissue%20%22jira_issue_key%3DJENKINS-47760%22">JENKINS-47760</a>]</i>
